File: src/main/java/org/angry1980/security/jwt/JwtExceptionTranslationWebFilter.java
package org.angry1980.security.jwt;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.authentication.AuthenticationCredentialsNotFoundException;
import org.springframework.security.web.server.AuthenticationEntryPoint;
import org.springframework.security.web.server.authorization.AccessDeniedHandler;
import org.springframework.security.web.server.authorization.HttpStatusAccessDeniedHandler;
import org.springframework.web.server.ServerWebExchange;
import org.springframework.web.server.WebFilter;
import org.springframework.web.server.WebFilterChain;
import reactor.core.publisher.Mono;

import java.util.Objects;

/**
 * Handler of authorization errors
 */
public class JwtExceptionTranslationWebFilter implements WebFilter {

    private static Logger LOG = LoggerFactory.getLogger(JwtExceptionTranslationWebFilter.class);

    private final AuthenticationEntryPoint entryPoint;
    private final AccessDeniedHandler accessDeniedHandler;

    public JwtExceptionTranslationWebFilter(AuthenticationEntryPoint entryPoint) {
        this(entryPoint, new HttpStatusAccessDeniedHandler(HttpStatus.FORBIDDEN));
    }

    public JwtExceptionTranslationWebFilter(AuthenticationEntryPoint entryPoint, AccessDeniedHandler accessDeniedHandler) {
        this.entryPoint = Objects.requireNonNull(entryPoint);
        this.accessDeniedHandler = Objects.requireNonNull(accessDeniedHandler);
    }

    @Override
    public Mono<Void> filter(ServerWebExchange exchange, WebFilterChain chain) {
        return chain.filter(exchange)
                .onErrorResume(AccessDeniedException.class, e ->
                    exchange.getPrincipal()
                            // user was not authenticated
                            .switchIfEmpty(
                                    Mono.defer(
                                            () -> entryPoint.commence(exchange, new AuthenticationCredentialsNotFoundException("Not Authenticated", e))
                                    )
                            )
                            // user is authenticated but does not have enough rights to get access to resource
                            .flatMap( principal -> {
                                LOG.info("Access for {} to {} is denied", principal.getName(), JwtUtils.getRequestInfo(exchange));
                                return accessDeniedHandler.<Void>handle(exchange, e);
                            })
                );
    }
}

File: src/main/java/org/angry1980/security/jwt/JwtExceptionTranslationWebFilter.kt
package org.angry1980.security.jwt

import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.authentication.AuthenticationCredentialsNotFoundException
import org.springframework.security.web.server.AuthenticationEntryPoint
import org.springframework.security.web.server.authorization.AccessDeniedHandler
import org.springframework.security.web.server.authorization.HttpStatusAccessDeniedHandler
import org.springframework.web.server.ServerWebExchange
import org.springframework.web.server.WebFilter
import org.springframework.web.server.WebFilterChain
import reactor.core.publisher.Mono
import java.security.Principal

/**
 * Handler of authorization errors
 */
class JwtExceptionTranslationWebFilter (val entryPoint: AuthenticationEntryPoint,
                                        val accessDeniedHandler: AccessDeniedHandler = HttpStatusAccessDeniedHandler(HttpStatus.FORBIDDEN)) : WebFilter {

    val LOG = LoggerFactory.getLogger(JwtExceptionTranslationWebFilter::class.java)

    override fun filter(exchange: ServerWebExchange, chain: WebFilterChain) = chain.filter(exchange)
            .onErrorResume(AccessDeniedException::class.java) { e ->
                exchange.getPrincipal<Principal>()
                        // user was not authenticated
                        .switchIfEmpty(
                                Mono.defer { entryPoint.commence<Principal>(exchange, AuthenticationCredentialsNotFoundException("Not Authenticated", e)) }
                        )
                        // user is authenticated but does not have enough rights to get access to resource
                        .flatMap {
                            LOG.info("Access for {} to {} is denied", it.getName(), exchange.requestInfo())
                            accessDeniedHandler.handle<Void>(exchange, e)
                        }

            }

}
